bundle/edu.gemini.p1monitor/src/main/scala/edu/gemini/p1monitor/P1MonitorDirMonitor.scala

```diff
@@ -36,22 +34,24 @@ class P1MonitorDirMonitor(cfg: P1MonitorConfig) extends DirListener {
       originalsDir.mkdirs()
     }
 
-    def writeUpdated(attachment: File, proposal: Proposal, f: File) {
+    def writeUpdated(attachment: File, proposal: Proposal, f: File):Proposal = {
       // Use just the name to avoid hardcoding the file
       val attachedFileName = attachment.getName
       LOG.info(s"Attachment file changed to $attachedFileName")
-      ProposalIo.write(proposal.copy(meta = proposal.meta.copy(attachment = Some(new File(attachedFileName)))), f)
+      val p = proposal.copy(meta = proposal.meta.copy(attachment = Some(new File(attachedFileName))))
+      ProposalIo.write(p, f)
+      p
     }
 
-    def updateAttachmentLink(newPDFile: Option[File], f: File) = {
+    def updateAttachmentLink(newPDFile: Option[File], f: File):Option[Proposal] = {
       // If we have a PDF we must replace the attachment name
-      newPDFile.foreach {
+      newPDFile.flatMap {
         case p =>
           // This sounds like a good idea, let Proposal to do the IO, but this will need some help to avoid writing
           // the absolute path of the file
-          for {
+          (for {
             conversion <- ProposalIo.readAndConvert(f)
-          } yield writeUpdated(p, conversion.proposal, f)
+          } yield writeUpdated(p, conversion.proposal, f)).toOption
```

Comment:
Not that I haven't done it anyway, but I always feel a bit sheepish about doing side-effects in a map.  Can the actual writing to a file be deferred so that the map is pure?